### PR TITLE
Don't alloc IEnumerable in Reset for ITlsConnectionFeature

### PR DESCRIPTION
**Pull request by benaadams**
Fixes https://github.com/aspnet/KestrelHttpServer/issues/2008

/cc @davidfowl @halter73

### CODE REVIEW

**Comment by davidfowl** [comment]
My fear is that this won't scale. We want to flow features from the connection to the request.

**Comment by davidfowl** [comment]
Actually it might be possible to just delegate instead of copy per request.

**Comment by benaadams** [comment]
Not sure what you mean? It it because there will be more than one `ITlsConnectionFeature` in the `ConnectionFeatures` and the lookup will return first?

So you need to enumerate every feature to pickup the last?

**Comment by davidfowl** [comment]
@benaadams features from the connection's feature collection should be accessible from the request's feature collection.

**Comment by benaadams** [comment]
I don't understand why enumerating would "scale" better than a look up though?

**Comment by benaadams** [comment]
Is it the stacking of FeatureCollections and you want to get to the bottom one?

**Comment by davidfowl** [comment]
> I don't understand why enumerating would "scale" better than a look up though?

https://github.com/aspnet/KestrelHttpServer/pull/2009#issuecomment-323650306

That's why I said it would probably be fine to just delegate instead of copy.

**Comment by benaadams** [comment]
You just linked to earlier comment that suggested an alternative implementation; but didn't provide a more information on the issue.

This change is doing the same thing as the enumerator; i.e. using the `ConnectionFeatures` to look up and set the `ITlsConnectionFeature`.

Where I could see the behavior differing is when the `ConnectionFeatures` is multiple stacked `FeatureCollection`s; so the enumerator will keep going until it gets to the bottom `FeatureCollection` that has a `ITlsConnectionFeature` (and keep going until its enumerated everything) whereas the lookup will use the `ITlsConnectionFeature` on the top `FeatureCollection`.

So is that the issue?

**Comment by davidfowl** [comment]
My initial reaction was that this was a problem because it was no longer copying features from the connection to the request feature collection. That's what I meant by "it won't scale". I realized that it would probably be better to delegate to the ConnectionFeatures itself rather than copy them to the request features on a per request basis.

TL;DR: This change is fine 😄

**Comment by benaadams** [comment]
Ahh... By delegate, just set to null and null ?? coalese in getter?

**Comment by davidfowl** [comment]
> Ahh... By delegate, just set to null and null ?? coalese in getter?

I meant fallback to ConnectionFeatures when asking for request features if not found in the current feature collection.

Frame.Features.cs

```C#
IFeatureCollection.this[Type key]
{
   get => return GetFeature(key) ?? ConnectionFeatures[key];
}
```

**Comment by benaadams** [comment]
Then can just null it an pay cost on lookup rather than for every Reset. Makes sense.

**Comment by benaadams** [comment]
Its an expensive way to set scheme for https though; will have a play

**Comment by davidfowl** [comment]
> Its an expensive way to set scheme for https though; will have a play

I think special casing the scheme like this is fine.

**Comment by benaadams** [comment]
I think I made it unhappy

File: src/Kestrel.Core/Internal/Http2/Http2Stream.FeatureCollection.cs [hunk 1]
```diff
@@ -246,13 +246,13 @@ MinDataRate IHttpMinResponseDataRateFeature.MinDataRate
 
         object IFeatureCollection.this[Type key]
         {
-            get => FastFeatureGet(key);
```
Comment (Tratcher) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.FeatureCollection.cs]: Beware of exposing the underlying features directly. Any mutable fields should be reset per request.

Comment (benaadams) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.FeatureCollection.cs]: They are reset further up by `ResetFeatureCollection();` in `Reset`

Comment (Tratcher) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.FeatureCollection.cs]: That resets the presence of a feature, not the values on that feature.

Comment (benaadams) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.FeatureCollection.cs]: Ah, like if you change `ClientCertificate` on `ITlsConnectionFeature`

Comment (halter73) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.FeatureCollection.cs]: I'm not so sure connection-level features should be reset be request. Then again, I'm not sure why you would want to mutate a connection-level feature via the request's feature collection. If there was a way to make these features immutable when accessed this way, we probably would.

File: src/Kestrel.Core/Internal/Http/Frame.cs [hunk 2]
```diff
@@ -362,19 +362,12 @@ public void Reset()
...
-                    if (feature.Key == typeof(ITlsConnectionFeature))
-                    {
-                        Scheme = "https";
-                    }
-
-                    FastFeatureSet(feature.Key, feature.Value);
-                }
+                var tlsFeature = ConnectionFeatures?[typeof(ITlsConnectionFeature)];
+                _scheme = tlsFeature != null ? "https" : "http";
             }
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http/Frame.cs]: Nit: newline.

Comment (benaadams) [inline src/Kestrel.Core/Internal/Http/Frame.cs]: Added

**Comment by halter73** [comment]
Thanks!

File: src/Kestrel.Core/Internal/Http2/Http2Stream.FeatureCollection.cs (hunk 1 above)
Comment (Tratcher) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.FeatureCollection.cs]: Will remote IP be a connection level feature? We mutate that all the time using x-forwarded-for. It should be reset per request as each request may have different forwarders.

So far I think you're handling the IPs correctly, copying them when you create the frame, but I'm less sure about the other features and their mutable fields.

Comment (Tratcher) [inline src/Kestrel.Core/Internal/Http2/Http2Stream.FeatureCollection.cs]: Yes, client cert is a good example.
